src/Pathy/Path.purs

```diff
@@ -0,0 +1,291 @@
+
+-- | A type describing a file or directory path.
+type AnyPath a = Either (Path a Dir) (Path a File)
+
+-- | A type describing a relative file or directory path.
+type RelPath = AnyPath Rel
+
+-- | A type describing an absolute file or directory path.
+type AbsPath = AnyPath Abs
+
+-- | A type describing a directory whose location is given relative to some
+-- | other, unspecified directory (referred to as the "current directory").
+type RelDir = Path Rel Dir
+
+-- | A type describing a directory whose location is absolutely specified.
+type AbsDir = Path Abs Dir
+
+-- | A type describing a file whose location is given relative to some other,
+-- | unspecified directory (referred to as the "current directory").
+type RelFile = Path Rel File
+
+-- | A type describing a file whose location is absolutely specified.
+type AbsFile = Path Abs File
+
+-- | The root directory, which can be used to define absolutely-located resources.
+rootDir :: Path Abs Dir
+rootDir = Init
+
+-- | The "current directory", which can be used to define relatively-located
+-- | resources.
+currentDir :: Path Rel Dir
+currentDir = Init
+
+-- | Creates a path which points to a relative file of the specified name.
+file :: NonEmptyString -> Path Rel File
+file = file' <<< Name
+
+-- | Creates a path which points to a relative file of the specified name.
+file' :: Name File -> Path Rel File
+file' = In currentDir
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir :: NonEmptyString -> Path Rel Dir
+dir = dir' <<< Name
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir' :: Name Dir -> Path Rel Dir
+dir' = In currentDir
+
+-- | Creates a path that points to the parent directory of the specified path.
+parentOf :: forall a. Path a Dir -> Path a Dir
+parentOf p = ParentOf p
+
+-- | Extends a path with a file or directory under the current path.
+extendPath :: forall a b. Path a Dir -> Name b -> Path a b
+extendPath p = In p
+
+-- | Given a directory path, appends a relative path to extend the original
+-- | path.
+appendPath :: forall a b. Path a Dir -> Path Rel b -> Path a b
+appendPath = case _, _ of
+  Init, Init -> Init
+  ParentOf p, Init -> ParentOf (p </> Init)
+  In p (Name d), Init -> In (p </> Init) (Name d)
+  p1, ParentOf p2 -> ParentOf (p1 </> p2)
+  p1, In p2 n -> In (p1 </> p2) n
+
+infixl 6 appendPath as </>
+
+-- | Ascends into the parent of the specified directory, then descends into
+-- | the specified path.
+-- |
+-- | ```purescript
+-- | canonicalize (rootDir </> dir "foo" <..> dir "bar") = rootDir </> dir "bar"
+-- | ```
+parentAppend :: forall a b. Path a Dir -> Path Rel b -> Path a b
+parentAppend d p = parentOf d </> p
+
+infixl 6 parentAppend as <..>
+
+-- | Canonicalizes a path, by reducing things in the form `/x/../` to just
+-- | `/x/`. Paths like `/../` will be normalized to `/`.
```

Comment:
I'm not sure that calling `/../` an invalid path is right really - it's not one that bash rejects anyway, it just gets treated as `/` if you try to use it. I guess we can drop the comment about it being invalid from the `Path` type too.

I guess it's like `././` - it's just something that gets normalized out rather than strictly being invalid.

Reply:
hm yes `$ cd /../../`, in that case we can just remove that as for print this case is normalized right

Reply:
That's true, yeah. `debugPrintPath` will show it in a non-canonicalized path still though, if that matters.

Reply:
Let's leave this as is, I think comment on Path type is fine too.